rudder-core/src/main/scala/com/normation/rudder/reports/execution/ReportsExecutionRepositoryImpl.scala

```diff
@@ -0,0 +1,167 @@
+* Copyright 2013 Normation SAS
+*************************************************************************************
+*
+* This file is part of Rudder.
+*
+* Rudder is free software: you can redistribute it and/or modify
+* it under the terms of the GNU General Public License as published by
+* the Free Software Foundation, either version 3 of the License, or
+* (at your option) any later version.
+*
+* In accordance with the terms of section 7 (7. Additional Terms.) of
+* the GNU General Public License version 3, the copyright holders add
+* the following Additional permissions:
+* Notwithstanding to the terms of section 5 (5. Conveying Modified Source
+* Versions) and 6 (6. Conveying Non-Source Forms.) of the GNU General
+* Public License version 3, when you create a Related Module, this
+* Related Module is not considered as a part of the work and may be
+* distributed under the license agreement of your choice.
+* A "Related Module" means a set of sources files including their
+* documentation that, without modification of the Source Code, enables
+* supplementary functions or services in addition to those offered by
+* the Software.
+*
+* Rudder is distributed in the hope that it will be useful,
+* but WITHOUT ANY WARRANTY; without even the implied warranty of
+* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+* GNU General Public License for more details.
+*
+* You should have received a copy of the GNU General Public License
+* along with Rudder.  If not, see <http://www.gnu.org/licenses/>.
+
+*
+*************************************************************************************
+*/
+
+package com.normation.rudder.reports.execution
+
+
+import com.normation.inventory.domain.NodeId
+import com.normation.rudder.db.DB
+import com.normation.rudder.repository.jdbc.PostgresqlInClause
+
+import net.liftweb.common._
+import com.normation.rudder.db.Doobie
+import scalaz.{Failure => _, _}, Scalaz._
+import doobie.imports._
+import scalaz.concurrent.Task
+
+case class RoReportsExecutionRepositoryImpl (
+    db: Doobie
+  , pgInClause : PostgresqlInClause
+) extends RoReportsExecutionRepository with Loggable {
+
+  import db._
+
+  override def getNodesLastRun(nodeIds: Set[NodeId]): Box[Map[NodeId, Option[AgentRun]]] = {
+    nodeIds.map( _.value).toList.toNel match {
+      case None => Full(Map())
+      case Some(nodes) =>
+
+        //val nodes = pgInClause.in("nodeid", nodeIds.map(_.value))
+        implicit val nodesParam = Param.many(nodes)
+
+        //notice the use of # ({} is forbidden) in place of ${} to interpolate the actual value of in,
+        //not make a sql parameter to the query
+        val sql = sql"""select distinct on (nodeid)
+                            nodeid, date, nodeconfigid, complete, insertionid
+                        from  reportsexecution
+                        where complete = true and nodeid in (${nodes : nodes.type})
```

Comment:
It is what the documentation says to do: https://tpolecat.github.io/doobie-0.2.4/05-Parameterized.html#dealing-with-in-clauses
And it's the type of that instance, because in scala, each class instance has its own type, that you can get like that, ex:
```scala
scala> val a = "plop"
a: String = plop
scala> {a: a.type}
res0: a.type = plop
}

I believe it is needed to get the correct inference so that the correct type class is selected to instantiate the correct "in" parameter.